toolkit/tools/pkg/imagecustomizerlib/customizeboot.go

```diff
@@ -107,143 +93,228 @@ func replaceToken(inputGrubCfgContent string, oldToken string, newToken string)
+				// There is a variable expansion (e.g. $a) in the arg.
+				// Stop here because we don't know what value to fill in for the variable.
+				hasVarExpansion = true
+				break subWordsLoop
+			}
+		}
+
+		argString := argStringBuilder.String()
+		name, value, foundEqSymbol := strings.Cut(argString, "=")
+		if !foundEqSymbol && hasVarExpansion {
+			// Arg has a variable expansion (e.g. $a) that is part of the arg name.
+			// There isn't any easy way to handle such args. So, just ignore them.
+			continue
+		}
+
+		if hasVarExpansion {
+			// The arg string value isn't known because it contains a variable expansion.
+			value = ""
+		}
+
+		arg := grubConfigLinuxArg{
+			Token:                argToken,
+			Name:                 name,
+			Value:                value,
+			ValueHasVarExpansion: hasVarExpansion,
+		}
+		args = append(args, arg)
 	}
 
-	if start == -1 {
-		return "", "", fmt.Errorf("failed to find kernel arg value for (%s)", name)
+	if insertAtToken == nil {
+		return nil, grub.Token{}, fmt.Errorf("failed to find $kernelopts in linux command line")
 	}
 
-	outputGrubCfgContent = fmt.Sprintf("%s%s=%s%s", inputGrubCfgContent[:start], name, value, inputGrubCfgContent[end:])
+	return args, *insertAtToken, nil
+}
 
-	return outputGrubCfgContent, oldValue, nil
+// Filters a list of kernel command-line args to only those that match the provided list of names.
+func findMatchingCommandLineArgs(args []grubConfigLinuxArg, names []string) []grubConfigLinuxArg {
+	matching := []grubConfigLinuxArg(nil)
+
+argsLoop:
+	for _, arg := range args {
+		for _, name := range names {
+			if arg.Name == name {
+				matching = append(matching, arg)
+				continue argsLoop
+			}
+		}
+	}
+
+	return matching
 }
 
-// Tries to remove the specified kernel CLI arg from the grub file. Does not fail if the arg is not present.
+// Tries to find the specified kernel CLI arg. Does not fail if the arg is not found.
 //
 // Returns:
-// - outputGrubCfgContent: The resulting grub.cfg file.
-// - index: The index in the grub.cfg where the arg was removed from. Or -1 if not found.
-func tryRemoveKernelCommandLineArgumentValue(inputGrubCfgContent string, name string) (string, int, error) {
-	// Search for the kernel CLI arg.
-	start, end, _, err := findKernelCommandLineArgumentArg(inputGrubCfgContent, name)
+// - value: The value of the arg. Or "" if not found.
+func findKernelCommandLineArgValue(args []grubConfigLinuxArg, name string) (string, error) {
+	foundArgs := findMatchingCommandLineArgs(args, []string{name})
+	if len(foundArgs) <= 0 {
+		return "", nil
+	}
+
+	lastArg := foundArgs[len(foundArgs)-1]
+	if lastArg.ValueHasVarExpansion {
+		return "", fmt.Errorf("kernel arg (%s) has variable expansion in value", name)
+	}
+
+	return lastArg.Value, nil
+}
+
+// Finds an existing kernel command-line arg and replaces its value.
+func replaceKernelCommandLineArgumentValue(inputGrubCfgContent string, name string, value string,
```

Comment:
nit: consistency Argument vs Arg in the function name

Reply:
Fixed.